Blog post whats new in Camel 3.7 (#508)
apache/camel-website: content/blog/2020/12/Camel37-Whatsnew/index.md

---
title: "Apache Camel 3.7 What's New"
date: 2020-12-17
authors: [davsclaus]
categories: ["Releases"]
preview: Details of what we have done in the Camel 3.7 release.
---

Apache Camel 3.7 LTS has just been released.

This is a LTS release which means we will provide patch releases for one year.
The next planned LTS release is 3.10 scheduled towards summer 2021.

## So what's in this release?

This release introduces a set of new features and noticeable improvements that we will cover in this blog post.

### Pre compiled languages

> We are continued our avenue of making Camel faster and smaller.

Reviewer: "We continued our avenue..." (remove "are")

components then this work will progress over the next couple of Camel releases.

### Separating Model and EIP processors 

In this release we untangled model, reifier and processors.

This is a great achievement which allows us to take this even further with design time vs runtime.

    Model    ->    Reifier   ->   Processor
    (startup)      (startup)      (runtime)

The model is the structure of the DSL which you can think of as _design time_ specifying your Camel routes.
The model is executed once during startup and via the reifier (factory) the runtime EIP processors is cretated.
After this work is done, the model is essentially not needed anymore.

By separating this into different JARs (camel-core-model, camel-core-reifier, camel-core-processor) then we ensure
they are separated and this allows us to better do built time optimizations and dead code elimination via Quarkus and GraalVM.

This brings up to lightweight mode.

### Lightweight mode

We started an experiment earlier with a lightweight mode. With the separation of the model from the processors,
then we have a great step forward, which allowed us to make the lightweight mode available for end users to turn on.

> In lightweight mode, Camel will after startup, then remove all references to the model, and other bits, that all together

Reviewer: Not clear, could you rephrase?

---

Apache Camel 3.7 LTS has just been released.

This is a LTS release which means we will provide patch releases for one year.
The next planned LTS release is 3.10 scheduled towards summer 2021.

## So what's in this release?

This release introduces a set of new features and noticeable improvements that we will cover in this blog post.

### Pre compiled languages

We continued our avenue of making Camel faster and smaller.
This time we focused on the built-in Simple scripting language.

First we added the [jOOR language](/components/latest/languages/joor-language.html). jOOR is a small Java tool
for performing runtime compilation of Java source code in-memory. It has some limitations but generally works well for small
scripting code (requires Java 11 onwards).

Then we worked on compiled simple.

### Compiled Simple

The [csimple language](https://camel.apache.org/components/latest/languages/csimple-language.html)
is parsed into regular Java source code and compiled together with all the other source code,
or compiled once during bootstrap via jOOR.

{{< image "csimple-compiled.png" "Compiled Simple Language" >}}

> In a nutshell where compiled simple excels over simple language is primary around dynamic OGNL method calls.

Reviewer 2: I had to read several times to understand this sentence.

What about something like:
In a nutshell, `Compiled Simple` excels primarily over simple language when using dynamic Object-Graph Navigation Language ([OGNL](https://en.wikipedia.org/wiki/OGNL) ) method calls..

and wtih a link to

For example profiling the following simple expression

    <simple>${exchangeProperty.user.getName} != null && ${exchangeProperty.user.getAge} > 11</simple>

with the equivalent csimple expression:

    <csimple>${exchangeProperty.user} != null && 
             ${exchangeProperty.user.getName()} != null && 
             ${exchangeProperty.user.getAge()} > 11</csimple>

yields a dramatic 100 times performance improvement in reduced cpu usage as shown in the screenshot:

{{< image "simple-csimple-performance.png" "Simple vs Compiled Simple" >}}

For more information about the compiled simple language and further break down of performance improvements
then read [Claus blog post](http://www.davsclaus.com/2020/12/apache-camel-37-compiled-simple.html).

> We have provided two small examples that demonstrate csimple as pre compiled and as runtime compiled during bootstrap.

Reviewer 2: ```suggestion
We have provided two small examples that demonstrate `csimple` as pre compiled and as runtime compiled during bootstrap.
```

If you are more curious about how we did these optimizations and with some performance numbers,
then [Claus wrote a blog post](http://www.davsclaus.com/2020/11/apache-camel-37-more-camel-core.html).

### Optimized components startup

The camel core has been optimized in Camel 3 to be small, slim, and fast on startup. This benefits Camel Quarkus which
can do built time optimizations that take advantage of the optimized camel core.

We have continued this effort in the Camel components where whenever possible initialization is moved ahead
to an earlier phase during startup, that allows enhanced built time optimizations. As there are a lot of Camel
components then this work will progress over the next couple of Camel releases.

### Separating Model and EIP processors 

In this release we untangled model, reifier and processors.

This is a great achievement which allows us to take this even further with design time vs runtime.

    Model    ->    Reifier   ->   Processor
    (startup)      (startup)      (runtime)

The model is the structure of the DSL which you can think of as _design time_ specifying your Camel routes.

> The model is executed once during startup and via the reifier (factory) the runtime EIP processors is cretated.

Reviewer 2: ```suggestion
The model is executed once during startup and via the reifier (factory) the runtime EIP processors is created.
```

After this work is done, the model is essentially not needed anymore.

By separating this into different JARs (camel-core-model, camel-core-reifier, camel-core-processor) then we ensure
they are separated and this allows us to better do built time optimizations and dead code elimination via Quarkus and GraalVM.

This brings up to lightweight mode.

### Lightweight mode

We started an experiment earlier with a lightweight mode. With the separation of the model from the processors,
then we have a great step forward, which allowed us to make the lightweight mode available for end users to turn on.

In lightweight mode Camel removes all references to the model after startup 
which causes the JVM to be able to garbage collect all model objects and unload classes, freeing up memory.

> After this it's no longer possible to dynamic add new Camel routes. The lightweight mode is intended for

Reviewer 2: ```suggestion
After this it's no longer possible to dynamically add new Camel routes. The lightweight mode is intended for
```

microservice/serverless architectures, with a _closed world_ assumption.

### Autowiring components

> The Camel components is now capable of autowiring by type. For example the AWS SQS components can automatic

Reviewer 2: ```suggestion
The Camel components is now capable of autowiring by type. For example the [AWS SQS](https://camel.apache.org/components/latest/aws-sqs-component.html) components can automatically
```

> lookup in the registry if there is single instance of `SqsClient`, and then pre configure itself.

Reviewer 2: ```suggestion
lookup in the registry if there is a single instance of `SqsClient`, and then pre configure itself.
```

We have marked up in the Camel documentation which component options supports this by showing *Autowired* in bold
in the description.

### Salesforce fixes

> Our recent Camel committer Jeremy Ross did great work to improve and fix bugs in the camel-salesforce component.

Reviewer 2: ```suggestion
Our recent Camel committer [Jeremy Ross](https://github.com/jeremyross) did great work to improve and fix bugs in the [camel-salesforce component](https://camel.apache.org/components/latest/salesforce-component.html).
```

We expect more to come from him.

### VertX Kafka Component

A new Kafka component has been developed that uses the Vert.X Kafka Java Client which allows us to use all of its
features, and also its robustness and stability. 

The [camel-vertx-kafka](https://camel.apache.org/components/latest/vertx-kafka-component.html) component is intended
to be (more) feature complete with the existing camel-kafka component. We will continue this work for the next couple of Camel releases.

The existing camel-kafka component has a few issues reported, which we want to fix in the upcoming releases,

> and at the same time we can offer this new camel-kafka-component as an alternative and potential a better component in the future.

Reviewer 2: you mean "we can offer this new camel-vertx-kafka component"?

### DataSonnet

The new [camel-datasonnet](https://camel.apache.org/components/latest/languages/datasonnet-language.html) component,

> is to be used for data transformation using the DataSonnet standard.

Reviewer 2: ```suggestion
is to be used for data transformation using the [DataSonnet](https://datasonnet.com/) standard.
```

DataSonnet is an open source JSON-centric, template-based data transformation standard built to rival
proprietary options available in the market.

### Spring Boot

We have upgraded to Spring Boot 2.4.

### New components

This release has a number of new components, data formats and languages:

> - AtlasMap: Transforms the message using an AtlasMap transformation

Reviewer 2: ```suggestion
- [AtlasMap](https://camel.apache.org/components/latest/atlasmap-component.html): Transforms the message using an [AtlasMap](https://www.atlasmap.io/) transformation
```
